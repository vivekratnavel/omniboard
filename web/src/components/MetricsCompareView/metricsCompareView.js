import React, {Component} from 'react';
import PropTypes from 'prop-types';
import './metricsCompareView.scss';
import Multiselect from 'react-bootstrap-multiselect';
import {Well, Alert, Button} from 'react-bootstrap';
import Select from 'react-select';
import backend from '../Backend/backend';
import {parseServerError, getAllPaths, getOption} from '../Helpers/utils';
import {MetricsPlotView} from '../MetricsPlotView/metricsPlotView';
import {ProgressWrapper} from '../Helpers/hoc';

class MetricsCompareView extends Component {
  static propTypes = {
    runIds: PropTypes.arrayOf(Number).isRequired,
    isSelected: PropTypes.bool.isRequired
  };

  runIdOptionsDomNode = null;

  constructor(props) {
    super(props);
    this.state = {
      isLoadingRuns: false,
      metrics: [],
      runIdOptions: [],
      error: '',
      metricLabels: ['_id'],
      metricLabelOptions: []
    };
  }

  _initializeState = () => {
    this._loadData();
    this._populateRunIdOptions();
  };

  componentDidMount() {
    const {isSelected} = this.props;
    if (isSelected) {
      // Load data only when the component is being displayed
      this._initializeState();
    }
  }

  componentDidUpdate(prevProps, _prevState, _snapshot) {
    // Populate runId options every time runIds change
    const {runIds, isSelected} = this.props;
    if (JSON.stringify(prevProps.runIds) !== JSON.stringify(runIds) && isSelected) {
      this._initializeState();
    }
  }

  _populateRunIdOptions = () => {
    const {runIds} = this.props;
    const runIdOptions = runIds.map(runId => {
      return {
        label: runId,
        value: runId,
        selected: true
      };
    });
    this.setState({
      runIdOptions
    });
  };

  _getSelectedRunIds = runIdOptions => runIdOptions.filter(option => option.selected === true)
    .map(option => Number(option.value));

  _handleRunIdsChange = _e => {
    const runIds = this.runIdOptionsDomNode.$multiselect.val();
    this.setState(({runIdOptions}) => ({
      runIdOptions: runIdOptions.map(option => {
        option.selected = runIds.includes(String(option.value));
        return option;
      })
    }));
  };

  _loadData = () => {
    const {runIds} = this.props;
    if (runIds.length >= 2) {
      this.setState({
        isLoadingRuns: true,
        error: ''
      });
      const queryString = JSON.stringify({
        run_id: {
          $in: runIds
        }
      });
      // Populate run with only required columns
      backend.get('api/v1/Metrics', {
        params: {
          query: queryString,
          populate: {
            path: 'run',
            select: ['-captured_out',
              '-meta',
              '-heartbeat',
              '-result',
              '-start_time',
              '-stop_time',
              '-resources',
              '-artifacts',
              '-format',
              '-command',
              '-info.metrics',
              '-experiment.sources']}
        }
      }).then(response => {
        // Recursively get all paths for populating select options for metric label
        const paths = response.data.reduce((acc, current) => {
          const keys = getAllPaths('', current.run[0], true);
          return [...new Set([...acc, ...keys])];
        }, []);
        const metricLabelOptions = paths.map(value => ({
          label: value,
          value
        }));
        this.setState({
          metrics: response.data,
          metricLabelOptions,
          isLoadingRuns: false
        });
      }).catch(error => {
        const message = parseServerError(error);
        this.setState({
          isLoadingRuns: false,
          error: message
        });
      });
    } else {
      this.setState({
        error: 'At-least two runs should be selected for comparison'
      });
    }
  };

  _handleMetricLabelChange = index => ({value}) => {
    this.setState(({metricLabels}) => {
      const metricLabelsClone = [...metricLabels];
      metricLabelsClone[index] = value;
      return {metricLabels: metricLabelsClone};
    });
  };

  _handleAddLabel = () => {
    // Add run id as default metric label
    this.setState(({metricLabels}) => ({
      metricLabels: [...metricLabels, '_id']
    }));
  };

  _handleDeleteLabel = index => () => {
    this.setState(({metricLabels}) => {
      const metricLabelsClone = [...metricLabels].filter((_label, i) => index !== i);
      return {metricLabels: metricLabelsClone};
    });
  };

  render() {
    const {metrics, isLoadingRuns, runIdOptions, error, metricLabels, metricLabelOptions} = this.state;
    const selectedRunIds = this._getSelectedRunIds(runIdOptions);
    const metricsResponseForPlot = metrics.filter(metric => selectedRunIds.includes(metric.run_id));
    const errorAlert = error ? <Alert bsStyle='danger'>{error}</Alert> : '';

    const metricLabelsDom = metricLabels.map((label, index) => {
      const selectTestAttr = `select-metric-label-${index}`;
      // Include separator for all selects except the last one
      const separator = metricLabels.length - 1 === index ? '' : '-';
      // Add delete button for all selects except the first one
      const deleteTestAttr = `delete-label-btn-${index}`;
      const deleteButton = index === 0 ? '' : (
        <Button className='minus-button' test-attr={deleteTestAttr} bsStyle='danger' bsSize='small' onClick={this._handleDeleteLabel(index)}>
          <span className='glyphicon glyphicon-minus' aria-hidden='true'/>
        </Button>
      );
      return (
        <span key={'metric-label-' + index}>
          <div className='metric-label-options-wrapper'>
            <Select
              className='select-metric-label'
              test-attr={selectTestAttr}
              options={metricLabelOptions}
              isLoading={isLoadingRuns}
              placeholder='Metric Label'
              value={getOption(label, metricLabelOptions)}
              onChange={this._handleMetricLabelChange(index)}
            />
          </div>
          {deleteButton}
          {separator}
        </span>
      );
    });
    return (
      <div className='metrics-compare-view'>
        <ProgressWrapper id='metrics-compare-progress-wrapper' loading={isLoadingRuns}>
          <div>
            {errorAlert}
            <Well bsSize='small' className='run-id-filter-well'>
              <h5>Runs: </h5>
              <div className='run-id-options-wrapper'>
                <Multiselect
                  ref={el => this.runIdOptionsDomNode = el}
                  includeSelectAllOption multiple
                  enableHTML
                  buttonText={(options, _select) => {
                    if (options.length === 0) {
                      return 'None selected';
                    }

                    return `${options.length} selected`;
                  }}
                  selectedClass='run-id-selected'
                  id='run_ids'
                  maxHeight={300}
                  data={runIdOptions}
                  onSelectAll={this._handleRunIdsChange}
                  onChange={this._handleRunIdsChange}
                  onDeselectAll={this._handleRunIdsChange}
                />
              </div>

              <h5>Metric Label: </h5>
              {metricLabelsDom}
              <div className='plus-button'>
                <Button test-attr='add-label-btn' bsStyle='info' bsSize='small' onClick={this._handleAddLabel}>
                  <span className='glyphicon glyphicon-plus' aria-hidden='true'/>
                </Button>
              </div>
            </Well>
            <MetricsPlotView metricsResponse={metricsResponseForPlot} runId='' metricLabels={metricLabels}/>
          </div>
        </ProgressWrapper>
      </div>
    );
  }
}

export {MetricsCompareView};
