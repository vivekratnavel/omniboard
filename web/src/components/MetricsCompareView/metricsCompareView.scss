
.run-id-filter-well {
  h5 {
    display: inline-block;
    margin-left: 10px;
  }
  .run-id-options-wrapper {
    display: inline-block;
    margin-left: 10px;
  }
  .metric-label-options-wrapper {
    margin-left: 10px;
    margin-right: 10px;
    width: 200px;
    display: inline-block;
  }
  .plus-button {
    margin-left: 10px;
    width: 70px;
    display: inline-block;
  }
  .minus-button {
    margin-left: -10px;
    margin-right: 10px;
  }
}