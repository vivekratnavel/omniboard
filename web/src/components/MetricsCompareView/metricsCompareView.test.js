import React from 'react';
import mockAxios from 'jest-mock-axios';
import {MetricsCompareView} from './metricsCompareView';

describe('MetricsCompareView', () => {
  let wrapper = null;
  const responseData = [
    {_id: '5a2b5a8c9c7a505a652f6127', name: 'pretrain.train.loss', run_id: 1, steps: [0, 1, 2, 3, 4],
      values: [0.7159541824544438, 0.3840367944955761, 0.3469185283233073, 0.30483262065173106, 0.28915774130337507],
      timestamps: ['2017-12-09T03:37:44.425Z', '2017-12-09T03:41:54.414Z', '2017-12-09T03:46:01.766Z', '2017-12-09T03:50:07.365Z', '2017-12-09T03:54:12.560Z'],
      run: [{_id: 1, experiment: {name: 'hello_config', base_dir: '/Users/Documents/sacred_experiment', dependencies: ['CairoSVG==2.4.1', 'sacred==0.7.5', 'svgwrite==1.3.1'], repositories: [], mainfile: 'hello2.py'}, host: {hostname: 'imac.lan', os: ['Darwin'], python_version: '3.7.7', cpu: 'AMD Ryzen 5', ENV: {}}, config: {message: 'Hello world!', recipient: 'world', seed: 797676031, tags: ['test1', 'test2'], train: {batch_size: 50, epochs: 68, lr: 0.01, settings: {epochs: 43}}}, status: 'COMPLETED', info: {}}]},
    {_id: '5a2b5aa09c7a505a652f6146', name: 'pretrain.val.loss', run_id: 2, steps: [0, 1, 2, 3, 4],
      values: [0.32177006650114165, 0.23237958704995795, 0.23340759051386187, 0.21925230575196739, 0.20541178824900605],
      timestamps: ['2017-12-09T03:38:01.945Z', '2017-12-09T03:42:11.673Z', '2017-12-09T03:46:18.843Z', '2017-12-09T03:50:24.377Z', '2017-12-09T03:54:29.752Z'],
      run: [{_id: 2, experiment: {name: 'hello_config', base_dir: '/Users/Documents/sacred_experiment', dependencies: ['CairoSVG==2.4.1', 'sacred==0.7.5', 'svgwrite==1.3.1'], repositories: [], mainfile: 'hello2.py'}, host: {hostname: 'imac.lan', os: ['Darwin'], python_version: '3.7.7', cpu: 'AMD Ryzen 5', ENV: {}}, config: {message: 'Hello world!', recipient: 'world', seed: 797676031, tags: ['test1', 'test2'], train: {batch_size: 50, epochs: 68, lr: 0.01, settings: {epochs: 43}}}, status: 'COMPLETED', info: {}}]},
    {_id: '5a2b5a8c9c7a505a652f6127', name: 'pretrain.train.loss', run_id: 3, steps: [0, 1, 2, 3, 4],
      values: [0.7859541824544438, 0.5640367944955761, 0.8769185283233073, 0.60483262065173106, 0.58915774130337507],
      timestamps: ['2017-12-09T03:37:44.425Z', '2017-12-09T03:41:54.414Z', '2017-12-09T03:46:01.766Z', '2017-12-09T03:50:07.365Z', '2017-12-09T03:54:12.560Z'],
      run: [{_id: 3, experiment: {name: 'hello_config', base_dir: '/Users/Documents/sacred_experiment', dependencies: ['CairoSVG==2.4.1', 'sacred==0.7.5', 'svgwrite==1.3.1'], repositories: [], mainfile: 'hello2.py'}, host: {hostname: 'imac.lan', os: ['Darwin'], python_version: '3.7.7', cpu: 'AMD Ryzen 5', ENV: {}}, config: {message: 'Hello world!', recipient: 'world', seed: 797676031, tags: ['test1', 'test2'], train: {batch_size: 50, epochs: 68, lr: 0.01, settings: {epochs: 43}}}, status: 'COMPLETED', info: {}}]},
    {_id: '5a2b5aa09c7a505a652f6146', name: 'pretrain.val.loss', run_id: 4, steps: [0, 1, 2, 3, 4],
      values: [0.52177006650114165, 0.9237958704995795, 0.11340759051386187, 0.41925230575196739, 0.90541178824900605],
      timestamps: ['2017-12-09T03:38:01.945Z', '2017-12-09T03:42:11.673Z', '2017-12-09T03:46:18.843Z', '2017-12-09T03:50:24.377Z', '2017-12-09T03:54:29.752Z'],
      run: [{_id: 4, experiment: {name: 'hello_config', base_dir: '/Users/Documents/sacred_experiment', dependencies: ['CairoSVG==2.4.1', 'sacred==0.7.5', 'svgwrite==1.3.1'], repositories: [], mainfile: 'hello2.py'}, host: {hostname: 'imac.lan', os: ['Darwin'], python_version: '3.7.7', cpu: 'AMD Ryzen 5', ENV: {}}, config: {message: 'Hello world!', recipient: 'world', seed: 797676031, tags: ['test1', 'test2'], train: {batch_size: 50, epochs: 68, lr: 0.01, settings: {epochs: 43}}}, status: 'COMPLETED', info: {}}]}
  ];
  const runIds = [1, 2, 3, 4];
  console.error = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <MetricsCompareView isSelected runIds={runIds}/>
    );
  });

  afterEach(() => {
    // Cleaning up the mess left behind the previous test
    mockAxios.reset();
    jest.clearAllMocks();
    wrapper = null;
  });

  describe('should load data', () => {
    it('success', async () => {
      expect(mockAxios.get.mock.calls).toHaveLength(1);

      expect(mockAxios.get.mock.calls[0][0]).toEqual('api/v1/Metrics');
      expect(wrapper.instance().state.isLoadingRuns).toBeTruthy();
      mockAxios.mockResponse({status: 200, data: responseData});

      await tick();
      expect(wrapper.update().instance().state.isLoadingRuns).toBeFalsy();
      expect(wrapper.update()).toMatchSnapshot();
    });

    it('error', () => {
      const errResponse = {status: 400, message: 'Not Found'};
      mockAxios.mockError(errResponse);

      expect(wrapper.instance().state.error).not.toEqual('');
    });
  });

  it('should not load data when runIds < 2', async () => {
    wrapper.unmount();
    mockAxios.reset();
    wrapper = mount(
      <MetricsCompareView isSelected runIds={[1]}/>
    );

    expect(mockAxios.get.mock.calls).toHaveLength(0);
    expect(wrapper.instance().state.error).not.toEqual('');
  });

  it('should handle run id change', async () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.reset();
    wrapper.instance().runIdOptionsDomNode = {
      $multiselect: {
        val: () => ['4', '3', '1']
      }
    };
    wrapper.instance()._handleRunIdsChange();

    expect(wrapper.instance().state.runIdOptions.find(option => option.value === 2).selected).toBeFalsy();
  });

  it('should handle metric label change', async () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.reset();
    expect(wrapper.instance().state.metricLabels[0]).toEqual('_id');
    wrapper.instance()._handleMetricLabelChange(0)({value: 'experiment.name'});

    expect(wrapper.instance().state.metricLabels[0]).toEqual('experiment.name');
  });

  it('should add new metric label', async () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.reset();
    expect(wrapper.instance().state.metricLabels).toHaveLength(1);
    wrapper.update().find('[test-attr="add-label-btn"]').at(0).simulate('click');

    expect(wrapper.instance().state.metricLabels).toHaveLength(2);
    expect(wrapper.instance().state.metricLabels[1]).toEqual('_id');
  });

  it('should delete a metric label', async () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.reset();
    const addButton = wrapper.update().find('[test-attr="add-label-btn"]').at(0);
    addButton.simulate('click');
    addButton.simulate('click');

    expect(wrapper.instance().state.metricLabels).toHaveLength(3);
    expect(wrapper.instance().state.metricLabels[2]).toEqual('_id');

    wrapper.instance()._handleMetricLabelChange(1)({value: 'experiment.name'});
    wrapper.instance()._handleMetricLabelChange(2)({value: 'host.hostname'});

    wrapper.update().find('[test-attr="delete-label-btn-1"]').at(0).simulate('click');
    expect(wrapper.instance().state.metricLabels).toHaveLength(2);
    expect(wrapper.instance().state.metricLabels[1]).toEqual('host.hostname');

    wrapper.update().find('[test-attr="delete-label-btn-1"]').at(0).simulate('click');
    expect(wrapper.instance().state.metricLabels).toHaveLength(1);
    expect(wrapper.instance().state.metricLabels[0]).toEqual('_id');
  });

  describe('componentDidUpdate', () => {
    beforeEach(() => {
      mockAxios.mockResponse({status: 200, data: responseData});
      mockAxios.reset();
    });

    it('loads data', async () => {
      wrapper.setProps({runIds: [1, 2]});

      expect(mockAxios.get.mock.calls).toHaveLength(1);
      expect(wrapper.instance().state.isLoadingRuns).toBeTruthy();
    });
  });
});
