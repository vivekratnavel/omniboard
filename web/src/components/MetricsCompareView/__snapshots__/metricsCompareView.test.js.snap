// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricsCompareView should load data success 1`] = `
<div
  className="metrics-compare-view"
>
  <ProgressWrapper
    id="metrics-compare-progress-wrapper"
    loading={false}
  >
    <div>
      <Well
        bsClass="well"
        bsSize="small"
        className="run-id-filter-well"
      >
        <h5>
          Runs: 
        </h5>
        <div
          className="run-id-options-wrapper"
        >
          <MultiSelect
            buttonText={[Function]}
            data={
              Array [
                Object {
                  "label": 1,
                  "selected": true,
                  "value": 1,
                },
                Object {
                  "label": 2,
                  "selected": true,
                  "value": 2,
                },
                Object {
                  "label": 3,
                  "selected": true,
                  "value": 3,
                },
                Object {
                  "label": 4,
                  "selected": true,
                  "value": 4,
                },
              ]
            }
            enableHTML={true}
            id="run_ids"
            includeSelectAllOption={true}
            maxHeight={300}
            multiple={true}
            onChange={[Function]}
            onDeselectAll={[Function]}
            onSelectAll={[Function]}
            selectedClass="run-id-selected"
          />
        </div>
        <h5>
          Metric Label: 
        </h5>
        <span
          key="metric-label-0"
        >
          <div
            className="metric-label-options-wrapper"
          >
            <StateManager
              className="select-metric-label"
              defaultInputValue=""
              defaultMenuIsOpen={false}
              defaultValue={null}
              isLoading={false}
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "label": "_id",
                    "value": "_id",
                  },
                  Object {
                    "label": "experiment.name",
                    "value": "experiment.name",
                  },
                  Object {
                    "label": "experiment.base_dir",
                    "value": "experiment.base_dir",
                  },
                  Object {
                    "label": "experiment.mainfile",
                    "value": "experiment.mainfile",
                  },
                  Object {
                    "label": "host.hostname",
                    "value": "host.hostname",
                  },
                  Object {
                    "label": "host.python_version",
                    "value": "host.python_version",
                  },
                  Object {
                    "label": "host.cpu",
                    "value": "host.cpu",
                  },
                  Object {
                    "label": "config.message",
                    "value": "config.message",
                  },
                  Object {
                    "label": "config.recipient",
                    "value": "config.recipient",
                  },
                  Object {
                    "label": "config.seed",
                    "value": "config.seed",
                  },
                  Object {
                    "label": "config.train.batch_size",
                    "value": "config.train.batch_size",
                  },
                  Object {
                    "label": "config.train.epochs",
                    "value": "config.train.epochs",
                  },
                  Object {
                    "label": "config.train.lr",
                    "value": "config.train.lr",
                  },
                  Object {
                    "label": "config.train.settings.epochs",
                    "value": "config.train.settings.epochs",
                  },
                  Object {
                    "label": "status",
                    "value": "status",
                  },
                ]
              }
              placeholder="Metric Label"
              test-attr="select-metric-label-0"
              value={
                Object {
                  "label": "_id",
                  "value": "_id",
                }
              }
            />
          </div>
        </span>
        <div
          className="plus-button"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsSize="small"
            bsStyle="info"
            disabled={false}
            onClick={[Function]}
            test-attr="add-label-btn"
          >
            <span
              aria-hidden="true"
              className="glyphicon glyphicon-plus"
            />
          </Button>
        </div>
      </Well>
      <MetricsPlotView
        metricLabels={
          Array [
            "_id",
          ]
        }
        metricsResponse={
          Array [
            Object {
              "_id": "5a2b5a8c9c7a505a652f6127",
              "name": "pretrain.train.loss",
              "run": Array [
                Object {
                  "_id": 1,
                  "config": Object {
                    "message": "Hello world!",
                    "recipient": "world",
                    "seed": 797676031,
                    "tags": Array [
                      "test1",
                      "test2",
                    ],
                    "train": Object {
                      "batch_size": 50,
                      "epochs": 68,
                      "lr": 0.01,
                      "settings": Object {
                        "epochs": 43,
                      },
                    },
                  },
                  "experiment": Object {
                    "base_dir": "/Users/Documents/sacred_experiment",
                    "dependencies": Array [
                      "CairoSVG==2.4.1",
                      "sacred==0.7.5",
                      "svgwrite==1.3.1",
                    ],
                    "mainfile": "hello2.py",
                    "name": "hello_config",
                    "repositories": Array [],
                  },
                  "host": Object {
                    "ENV": Object {},
                    "cpu": "AMD Ryzen 5",
                    "hostname": "imac.lan",
                    "os": Array [
                      "Darwin",
                    ],
                    "python_version": "3.7.7",
                  },
                  "info": Object {},
                  "status": "COMPLETED",
                },
              ],
              "run_id": 1,
              "steps": Array [
                0,
                1,
                2,
                3,
                4,
              ],
              "timestamps": Array [
                "2017-12-09T03:37:44.425Z",
                "2017-12-09T03:41:54.414Z",
                "2017-12-09T03:46:01.766Z",
                "2017-12-09T03:50:07.365Z",
                "2017-12-09T03:54:12.560Z",
              ],
              "values": Array [
                0.7159541824544438,
                0.3840367944955761,
                0.3469185283233073,
                0.30483262065173106,
                0.28915774130337507,
              ],
            },
            Object {
              "_id": "5a2b5aa09c7a505a652f6146",
              "name": "pretrain.val.loss",
              "run": Array [
                Object {
                  "_id": 2,
                  "config": Object {
                    "message": "Hello world!",
                    "recipient": "world",
                    "seed": 797676031,
                    "tags": Array [
                      "test1",
                      "test2",
                    ],
                    "train": Object {
                      "batch_size": 50,
                      "epochs": 68,
                      "lr": 0.01,
                      "settings": Object {
                        "epochs": 43,
                      },
                    },
                  },
                  "experiment": Object {
                    "base_dir": "/Users/Documents/sacred_experiment",
                    "dependencies": Array [
                      "CairoSVG==2.4.1",
                      "sacred==0.7.5",
                      "svgwrite==1.3.1",
                    ],
                    "mainfile": "hello2.py",
                    "name": "hello_config",
                    "repositories": Array [],
                  },
                  "host": Object {
                    "ENV": Object {},
                    "cpu": "AMD Ryzen 5",
                    "hostname": "imac.lan",
                    "os": Array [
                      "Darwin",
                    ],
                    "python_version": "3.7.7",
                  },
                  "info": Object {},
                  "status": "COMPLETED",
                },
              ],
              "run_id": 2,
              "steps": Array [
                0,
                1,
                2,
                3,
                4,
              ],
              "timestamps": Array [
                "2017-12-09T03:38:01.945Z",
                "2017-12-09T03:42:11.673Z",
                "2017-12-09T03:46:18.843Z",
                "2017-12-09T03:50:24.377Z",
                "2017-12-09T03:54:29.752Z",
              ],
              "values": Array [
                0.32177006650114165,
                0.23237958704995795,
                0.23340759051386187,
                0.21925230575196739,
                0.20541178824900605,
              ],
            },
            Object {
              "_id": "5a2b5a8c9c7a505a652f6127",
              "name": "pretrain.train.loss",
              "run": Array [
                Object {
                  "_id": 3,
                  "config": Object {
                    "message": "Hello world!",
                    "recipient": "world",
                    "seed": 797676031,
                    "tags": Array [
                      "test1",
                      "test2",
                    ],
                    "train": Object {
                      "batch_size": 50,
                      "epochs": 68,
                      "lr": 0.01,
                      "settings": Object {
                        "epochs": 43,
                      },
                    },
                  },
                  "experiment": Object {
                    "base_dir": "/Users/Documents/sacred_experiment",
                    "dependencies": Array [
                      "CairoSVG==2.4.1",
                      "sacred==0.7.5",
                      "svgwrite==1.3.1",
                    ],
                    "mainfile": "hello2.py",
                    "name": "hello_config",
                    "repositories": Array [],
                  },
                  "host": Object {
                    "ENV": Object {},
                    "cpu": "AMD Ryzen 5",
                    "hostname": "imac.lan",
                    "os": Array [
                      "Darwin",
                    ],
                    "python_version": "3.7.7",
                  },
                  "info": Object {},
                  "status": "COMPLETED",
                },
              ],
              "run_id": 3,
              "steps": Array [
                0,
                1,
                2,
                3,
                4,
              ],
              "timestamps": Array [
                "2017-12-09T03:37:44.425Z",
                "2017-12-09T03:41:54.414Z",
                "2017-12-09T03:46:01.766Z",
                "2017-12-09T03:50:07.365Z",
                "2017-12-09T03:54:12.560Z",
              ],
              "values": Array [
                0.7859541824544438,
                0.5640367944955761,
                0.8769185283233073,
                0.604832620651731,
                0.5891577413033751,
              ],
            },
            Object {
              "_id": "5a2b5aa09c7a505a652f6146",
              "name": "pretrain.val.loss",
              "run": Array [
                Object {
                  "_id": 4,
                  "config": Object {
                    "message": "Hello world!",
                    "recipient": "world",
                    "seed": 797676031,
                    "tags": Array [
                      "test1",
                      "test2",
                    ],
                    "train": Object {
                      "batch_size": 50,
                      "epochs": 68,
                      "lr": 0.01,
                      "settings": Object {
                        "epochs": 43,
                      },
                    },
                  },
                  "experiment": Object {
                    "base_dir": "/Users/Documents/sacred_experiment",
                    "dependencies": Array [
                      "CairoSVG==2.4.1",
                      "sacred==0.7.5",
                      "svgwrite==1.3.1",
                    ],
                    "mainfile": "hello2.py",
                    "name": "hello_config",
                    "repositories": Array [],
                  },
                  "host": Object {
                    "ENV": Object {},
                    "cpu": "AMD Ryzen 5",
                    "hostname": "imac.lan",
                    "os": Array [
                      "Darwin",
                    ],
                    "python_version": "3.7.7",
                  },
                  "info": Object {},
                  "status": "COMPLETED",
                },
              ],
              "run_id": 4,
              "steps": Array [
                0,
                1,
                2,
                3,
                4,
              ],
              "timestamps": Array [
                "2017-12-09T03:38:01.945Z",
                "2017-12-09T03:42:11.673Z",
                "2017-12-09T03:46:18.843Z",
                "2017-12-09T03:50:24.377Z",
                "2017-12-09T03:54:29.752Z",
              ],
              "values": Array [
                0.5217700665011417,
                0.9237958704995795,
                0.11340759051386187,
                0.41925230575196737,
                0.905411788249006,
              ],
            },
          ]
        }
        runId=""
        stateFilterKeys={
          Array [
            "selectedMetricNames",
            "selectedXAxis",
            "selectedYAxis",
            "plotWidth",
            "plotHeight",
            "smoothing",
            "plotModes",
          ]
        }
      />
    </div>
  </ProgressWrapper>
</div>
`;
